contracts/MarginCalculator.sol

```diff
@@ -2,11 +2,198 @@
+        require(_vault.shortOtokens.length <= 1, "MarginCalculator: Too many short otokens in the vault.");
+        require(_vault.longOtokens.length <= 1, "MarginCalculator: Too many long otokens in the vault.");
+        require(_vault.collateralAssets.length <= 1, "MarginCalculator: Too many collateral assets in the vault.");
+
+        require(
+            _vault.shortOtokens.length == _vault.shortAmounts.length,
+            "MarginCalculator: Short asset and amount mismatch."
+        );
+        require(
+            _vault.longOtokens.length == _vault.longAmounts.length,
+            "MarginCalculator: Long asset and amount mismatch."
+        );
+        require(
+            _vault.collateralAssets.length == _vault.collateralAmounts.length,
+            "MarginCalculator: Collateral asset and amount mismatch."
+        );
+
+        int256 collateralAmount = _vault.collateralAmounts.length > 0
+            ? FixedPointInt256.uintToInt(_vault.collateralAmounts[0])
+            : 0;
+        int256 shortAmount = _vault.shortAmounts.length > 0 ? FixedPointInt256.uintToInt(_vault.shortAmounts[0]) : 0;
+        int256 longAmount = _vault.longAmounts.length > 0 ? FixedPointInt256.uintToInt(_vault.longAmounts[0]) : 0;
+
+        // No short tokens: return collateral value.
+        if (_vault.shortOtokens.length == 0) return (collateralAmount.intToUint(), true);
+
+        address short = _vault.shortOtokens[0];
+
+        // For the currenct version, ensure denominated == short.collateral
+        require(
+            OtokenInterface(short).collateralAsset() == _demonimated,
+            "MarginCalculator: Denomintated token should be short.collateral"
+        );
+
+        bool expired = now > OtokenInterface(short).expiryTimestamp();
+        bool isPut = OtokenInterface(short).isPut();
+        int256 netOtoken = 0;
+        if (expired) {
+            address long = _vault.longOtokens.length > 0 ? _vault.longOtokens[0] : address(0);
+            int256 shortCashValue = FixedPointInt256.uintToInt(getExpiredCashValue(short));
+            int256 longCashValue = FixedPointInt256.uintToInt(getExpiredCashValue(long));
+
+            // Net otoken value = (long cash value * long amount) - (short cash value * short amount)
+            int256 netOtokenAfterExpiry = (longCashValue.mul(longAmount)).sub(shortCashValue.mul(shortAmount));
+
+            if (isPut) {
+                /*
+                 * Put otoken net after expiry: same as above:
+                 *     (long cash value * long amount) - (short cash value * short amount)
+                 */
+                netOtoken = netOtokenAfterExpiry;
+            } else {
+                /*
+                 * Call otoken net after expiry: (net otoken value / underlying price)
+                 */
+                (uint256 underlyingPrice, ) = _getUnerlyingPrice(address(short));
+                // todo: do we need to check isFinalized?
```

Comment:
To check if price is finalized, we can use the `isPriceFinalized()` from Oracle contract

Reply:
I'm not sure if we need to check or not.

Reply:
What is the expect behavior in this period, (after expiry, before finalized?)

Reply:
I think after expiry before finalized, user should not be able to do anything related to the short option, so I think we should revert if price not finalized ?

Reply:
Can we do that  (Ban minting and burning after expiry) in the controller? Because this is a view function, if we use revert, then nobody can query this function during that locking period.

Reply:
I've added the check. But worth thinking about this view function issue in terms of UX.